security/advancedtls/provider.go

```diff
@@ -0,0 +1,199 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package advancedtls
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"time"
+
+	"google.golang.org/grpc/credentials/tls/certprovider"
+	"google.golang.org/grpc/grpclog"
+)
+
+const (
+	defaultInterval = 1 * time.Hour
+)
+
+var logger = grpclog.Component("channelz")
+
+// IdentityPemFileProviderOptions contains fields to be filled out by users
+// for obtaining identity private key and certificates from PEM files.
+type IdentityPemFileProviderOptions struct {
+	CertFile string
+	KeyFile  string
+	Interval time.Duration
+}
+
+// IdentityPemFileProvider implements certprovider.Provider.
+// It provides the most up-to-date identity private key
+// and certificates based on the input PEM files.
+type IdentityPemFileProvider struct {
+	distributor *certprovider.Distributor
+	closeFunc   func()
+}
+
+// RootPemFileProviderOptions contains fields to be filled out by users
+// for obtaining root certificates from PEM files.
+type RootPemFileProviderOptions struct {
+	TrustFile string
+	Interval  time.Duration
+}
+
+// RootPemFileProvider implements certprovider.Provider.
+// It provides the most up-to-date root certificates based on the input PEM files.
+type RootPemFileProvider struct {
+	distributor *certprovider.Distributor
+	closeFunc   func()
+}
+
+// NewIdentityPemFileProvider uses IdentityPemFileProviderOptions to construct a IdentityPemFileProvider.
+func NewIdentityPemFileProvider(o *IdentityPemFileProviderOptions) (*IdentityPemFileProvider, error) {
+	if o.CertFile == "" {
+		return nil, fmt.Errorf("users need to specify certificate file path in NewIdentityPemFileProvider")
+	}
+	if o.KeyFile == "" {
+		return nil, fmt.Errorf("users need to specify key file path in NewIdentityPemFileProvider")
+	}
+	// If interval is not set by users explicitly, we will set it to default interval.
+	if o.Interval == 0*time.Nanosecond {
+		o.Interval = defaultInterval
+	}
+	ticker := time.NewTicker(o.Interval)
+	quit := make(chan bool)
+	provider := &IdentityPemFileProvider{}
+	provider.distributor = certprovider.NewDistributor()
+	// Initialize the distributor with an empty KeyMaterial.
+	provider.distributor.Set(&certprovider.KeyMaterial{}, nil)
```

Comment:
There is no need to initialize with empty key material. In fact, you should not be doing that. Not initializing with empty keyMaterial makes sure that any calls to `KeyMaterial()` will block, which is what you really want.

Reply:
The reason we initialize it with an empty keyMaterial is because when the initial files are empty, we want to have a valid temporary keyMaterial. Otherwise, it will be outputting an error when the initial file contents are empty.
Do you have any suggestion on how to handle this case?

Reply:
Cindy, if the initial files are empty, there is no way we can provide a valid key material. Even if you do this, calling tls.LoadX509KeyPair on empty files will still cause the error, and we'll still log the message. Let's just remove this line.
 The logging is mainly to inform users that they haven't provided any valid credentials yet. But if we have cases when users set credentials after starting several connections(and those connections will block), we might pollute the output logs if we log too much. @easwars Do you think we need this logging here?

Reply:
Sure, I'll delete that.